aerosense-ai/data-gateway
# Timestamp correction support

# Summary

Adds timestamp correction to inbound packets.

We've been having time synchronisation problems until today, which this PR solves whilst also avoiding race conditions related to the synchronisation which were happening in the packet parser (by adding a `packet_timestamp` to the packet metadata that gets put on the queue, and processing the timestamp updates in the parser thread, not the reader thread).
 
Packet timestamp offset is continuously updated (from the node timestamp in `Constat` packets) but the determined value has been jittery, almost certainly because of the overhead of computing `time.time()` in python on a CPU with variable capacity. So I'm using a median filter to stabilise that and its working smoothly with less than 1/1000s variance when tested on the RPi.
 
The offset (whose value incorporates the combined drift rate of RPi + Node clocks) does drift significantly: almost 500ppm (0.5ms/s). So, the question was "is it the Node or the RPi that's drifting?". Using chronyc (the unix NTP utility installed on the RPi host OS) suggests that the RPi clock is only 6.5ppm fast, and should be further compensated for that.
 
This is good news because it suggests that we're compensating correctly (correcting fast-drifting node timestamps to the slow drifting absolute timestamp on the RPi rather than the other way around).
 
With 6.5ppm drift rate, and a 4.5hr NTP refresh interval, maximum drift we should experience is 0.1s.
 
The median filter is able to handle a 0.1s jump back in time smoothly (eg when the NTP sync kicks in), meaning we shouldn't ever get to the situation where a subsequent sample is timestamped earlier than a previous one: i.e. we should never be time-travelling.
 


# Contents ([#86](https://github.com/aerosense-ai/data-gateway/pull/86))

### New features
- Add median filter to prevent timestamp jitter

### Enhancements
- Implement thread safe packet timestamp correction

### Fixes
- Pop index rather than remove matching element
- Make cli tests use constats packet

### Operations
- Update version for the new timestamp correction feature

### Refactoring
- Simply control loop and lose redundant sleep state variable
- Align variable names correctly

### Testing
- Add test to ensure no data saved if timestamp not retrievable
- Update test for new info messages

### Uncategorised!
- Merge branch 'main' into feature/multiple-node-support
